Fixing CI

madhavajay opened the pull request:
## Description
Fixing CI.

## Affected Dependencies
None

## How has this been tested?
Locally and with CI.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

madhavajay committed:
Touch

madhavajay committed:
Touch back

madhavajay committed:
Updated GitHub CI

- Ran black black-nb mypy
- Removed python 3.5
- Fixed issue with MacOS Build using --features=-supports_dynamic_linker
- Updated Google DP to the latest commit
- Renamed Bazel target from bindings_test to pydp
- Updated pyproject.toml
- Fixed com_google_protobuf issue in WORKSPACE
- Disabled docker build on every commit / push

madhavajay committed:
Added pre-commit yaml file

- Ran pre-commit to fix lots of code issues
- Removed unused # type: ignore
- Updated setup.cfg with isort, mypy and flake8 rules
- Ran isort
- Disabled notebook tests
- Fixed matrix.os template variable in tests.yml

madhavajay committed:
Update docker to use python 3.9

- Fix mypy issues with missing typeshed
- Fixing tests.yml

madhavajay committed:
Fixing Dockerfile path to bazelisk

- Fixing issue with converting matrix.os to bazel config name

madhavajay committed:
Add ubuntu to .bazelrc file

madhavajay committed:
Fixing matrix.os conversion to bazel config

madhavajay committed:
Fixing multi-os issues in tests.yml

madhavajay committed:
More multi-os fixes

madhavajay committed:
Fixing windows bazel

madhavajay committed:
Fix windows

madhavajay committed:
Adding bazel caching for linux

- Fixing install wheel on windows
- Fixing print version

madhavajay committed:
More CI fixes

chinmayshah99 commented:
Does this also solve #358 ?

madhavajay committed:
tunshell

madhavajay committed:
Fixing macos wheel

madhavajay committed:
remove tunshell

madhavajay commented:
> Does this also solve #358 ?

Kind of. So it will build and install in CI, but really we should make sure that the wheel it creates can also install on any version of MacOS. There might be some extra tricks to making that happen, and we will likely need to add a second entry to the matrix to be sure:

```
macOS Big Sur 11.0 | macos-11.0
macOS Catalina 10.15 | macos-latest or macos-10.15
```

madhavajay commented:
I also think this should be done a different way:
https://github.com/OpenMined/PyDP/blob/dev/tests/notebooks/test_all_notebooks.py

But before refactoring it would be good to get a clear understanding of the goal? I would highly suggest running linting and checks before committing in pre-commit and then again in the linting step of CI, but currently the pre-commit checks aren't all passing because there are a lot of flake8 issues.

Try this:
```
$ pre-commit run --all-files
Found 216 errors in 29 files (checked 42 source files)
```

However, I think they should be handled as a seperate set of tasks so for now the key is just to get the tests passing and the build working so that we can do a Python 3.9 release.

Once these are passing for 3.9 I can add a nightly build to run the older versions like 3.6, 3.7 and 3.8 every night, and a simple file which can be used to trigger them manually by doing something like this:
```
$ date > tests/trigger/versions
```

madhavajay commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,57 @@
+      - id: check-executables-have-shebangs
+      - id: debug-statements
+      - id: name-tests-test
+        exclude: ^(tests/syft/lib/allowlist_report.py)
```
remove

replomancer commented:
I only investigated `test_bounded_mean.py` There may be some issues with other tests (they fail for me). Here's what I found so far.

This test:
https://github.com/OpenMined/PyDP/blob/c387752d2b42aebdce52f8fe687d0d9aab828b45/tests/algorithms/test_bounded_mean.py#L69-L73
means if you compute a dp mean of many many many fives, the result should be close to five.

Apparently for performance reasons, instead of building big lists of fives, an attempt is made here to load data:
https://github.com/OpenMined/PyDP/blob/c387752d2b42aebdce52f8fe687d0d9aab828b45/tests/algorithms/test_bounded_mean.py#L49-L62

from here:
https://github.com/OpenMined/PyDP/blob/c387752d2b42aebdce52f8fe687d0d9aab828b45/tests/algorithms/test_bounded_mean/test_bounded_mean_int64_data.proto#L1-L2
This file looks a little different when something useful is serialized here. Currently instead of having many many fives, BoundedMean probably gets nothing and produces random results between lower and upper bounds. This test passes after data file is deleted but it takes almost 2 minutes on my machine.

~Speaking of lower and upper bounds. Maybe something builds differently for me or somehow I downloaded another C++ lib version but for me this is also broken:~ (this was a result of some dev changes I was missing)
https://github.com/OpenMined/PyDP/blob/c387752d2b42aebdce52f8fe687d0d9aab828b45/tests/algorithms/test_bounded_mean.py#L47
~i.e. I get~
```
E       RuntimeError: LInf sensitivity has to be positive but is 0
```
~because~
```python
        x = BoundedMean(1.0, 0, 0, 10, dtype="int64")
```
~means~
```python
        # lower_bound same as upper_bound:
        x = BoundedMean(epsilon=1.0, lower_bound=0, upper_bound=0, l0_sensitivity=10, dtype="int64")
```
~I also get the same sensitivity error from here:~
https://github.com/OpenMined/PyDP/blob/c387752d2b42aebdce52f8fe687d0d9aab828b45/tests/algorithms/test_bounded_mean.py#L77-L78
~I need to increase upper_bound for it to stop crashing.~

I may have ideas/comments tomorrow.

replomancer commented:
(edit: this was from missing some dev changes)
~Other fails I see also result from sensitivity set to zero (LInf sensitivity is computed from other parameters) e.g.~
https://github.com/OpenMined/PyDP/blob/c387752d2b42aebdce52f8fe687d0d9aab828b45/tests/algorithms/test_bounded_variance.py#L6-L11
~Here~
```python
BoundedVariance(epsilon, delta, lower_bound, upper_bound, dtype="float") 
```
~actually means~
```python
BoundedVariance(epsilon=epsilon, lower_bound=delta, upper_bound=lower_bound, l0_sensitivity=upper_bound, dtype="float") 
```
~so both `lower_bound` and `upper_bound` are set to `0`.~

~Similar story in `TestBoundedStandardDeviation` and `TestBoundedSum`.~

chinmayshah99 commented on .bazelrc:
```diff
@@ -1,3 +1,3 @@
-build:macOS --cxxopt='-std=c++17'
-build:Windows --cxxopt='/std:c++17'
+build:linux --cxxopt='-std=c++17'
+build:macos --cxxopt='-std=c++17' --features=-supports_dynamic_linker
```
Is there a reason why these flags are added to macOS only?

madhavajay replied:
Yes, because they are only needed for macOS

chinmayshah99 commented on .github/workflows/pypipublish_linux.yml:
```diff
@@ -54,7 +54,7 @@ jobs:
       - name: Build pydp lib
         run: |
-          bazel build --config Linux src/python:bindings_test  --verbose_failures
+          bazel build --config ${{ matrix.os }} src/python:pydp  --verbose_failures
```
I am not sure if `matrix.os` would pick OS since this is hardcoded. Is this verified?

madhavajay replied:
I have actually refactored these now that the tests pass anyway, so its going to be identical to the existing tests yaml which works.

replomancer committed:
Quickfix serialized test data

chinmayshah99 commented on build_PyDP.sh:
```diff
@@ -4,11 +4,11 @@
-  *"linux"*)
-    PLATFORM="Linux"
+    *"linux"*)
+        PLATFORM="Linux"
```
can we change this to: `PLATFORM="linux"`.
Needed due to change over here: https://github.com/OpenMined/PyDP/blob/c387752d2b42aebdce52f8fe687d0d9aab828b45/.bazelrc

madhavajay committed:
Added bazel-cache

- Combined publish actions into a single github action
- Added nightly tests for older versions of python
- Updated some package meta-data
- Bumped version to 1.0.3
- Added poetry.lock file to speed up CI

madhavajay committed:
Removed unused exclusion rule from pre-commit

madhavajay committed:
Adding other OSes to the Bazel Cache step

madhavajay committed:
Update test to include information about cached result from
tests/algorithms/test_bounded_mean/test_bounded_mean_int64_data.proto

madhavajay committed:
Refactored Docker and Docs actions

- Updated poetry lock file

madhavajay committed:
Renaming a few action names

- Testing versions cron

madhavajay committed:
tunshell into windows machine

madhavajay committed:
tunshell into mac

madhavajay committed:
Revert cron job for versions

- Remove tunshell for mac

madhavajay committed:
Trying specific version of cache plugin for windows tar cache error

madhavajay committed:
Added proper bazel cache directories for each OS plus Windows fix

madhavajay committed:
Trying different tar.exe fix for Windows

madhavajay committed:
Forgot to remove --output_base ./bazel-cache

madhavajay committed:
Trying windows tar fix

- Busting cache on linux and mac

madhavajay committed:
Trying older bazel cache strategy

madhavajay committed:
Debugging Windows Bazel Cache

madhavajay committed:
Trying actions/cache@v2.1.5

madhavajay committed:
Trying a different windows bazel cache approach

madhavajay committed:
Fix test yaml

- Trying Windows Bash tunshell

madhavajay committed:
woops checkout not cache

madhavajay committed:
Wrong runner

madhavajay committed:
Windows good times

madhavajay committed:
Windows cache might be working

madhavajay committed:
Try without symlinks?

madhavajay committed:
Putting cache code in all workflows

- Time to test Windows cache retrieval

madhavajay committed:
Added linting requirement back to main tests

chinmayshah99 commented on tests/algorithms/test_bounded_mean.py:
```diff
@@ -62,9 +68,12 @@ def _make_loaded_object(value, size, iter):
     x = make_loaded_object(5, 100000000, 5)
-    assert expect_near(5.0, x.result(), 0.1)
+    res = x.result()
+    print(x, type(x), res, type(res))
```
remove this print statement from here?

chinmayshah99 commented on tests/algorithms/test_bounded_mean/test_bounded_mean_int64_data.proto:
```diff
@@ -1,2 +1,2 @@
-
-;type.googleapis.com/differential_privacy.BoundedMeanSummary
+M
+;type.googleapis.com/differential_privacy.BoundedMeanSummary�ʵ���	
```
some extra characters are added to the file. Verify if this is really needed?

replomancer replied:
This is needed for `test_bounded_mean_int64`.

This file has a very confusing name. It's called `test_bounded_mean_int64_data.proto`, but it doesn't store a protobuf message type definition, instead it actually stores some serialized data. The reason Madhava's PR was not passing checks was because this thing was missing here. If you restore the old version of this file (remove weird characters), you will see `test_bounded_mean_int64` failing again.

chinmayshah99 commented on prereqs_linux.sh:
```diff
@@ -22,7 +22,7 @@ else
     sudo apt-get install python3.6
 fi
 
```
change the SHA on line 62

chinmayshah99 replied:
Also, this needs an update in prereqs_osx.sh

madhavajay committed:
Fixed PR change requests

madhavajay committed:
Replace bazel with bazelisk

- Use dp submodule SHA

madhavajay committed:
debugging versions check

chinmayshah99 reviewed (changes requested):
LGTM! can you remove the tunnel ssh?

madhavajay committed:
Adding python binary and site-packages for different versions on unix

madhavajay committed:
Test all versions on mac and linux

madhavajay committed:
Adding Windows support for python path detection

madhavajay committed:
Reversed quotes and removed \ line endings

madhavajay committed:
Changing the var name on windows

madhavajay committed:
Using old syntax for the PYTHONPATH on windows

madhavajay committed:
Re-enabled windows on versions

madhavajay committed:
Replacing incorrect Capital R in replace

madhavajay commented:
Versions less than 3.9 are now building correctly and all tests are passing: https://github.com/OpenMined/PyDP/actions/runs/875497191
